src/Temporalio.Extensions.Hosting/ServiceProviderExtensions.cs

```diff
@@ -55,27 +55,42 @@ public static class ServiceProviderExtensions
         /// <param name="method">Method to create activity definition from.</param>
         /// <returns>Created definition.</returns>
         public static ActivityDefinition CreateTemporalActivityDefinition(
-            this IServiceProvider provider, Type instanceType, MethodInfo method)
+            this IServiceProvider provider,
+            Type instanceType,
+            MethodInfo method)
         {
             // Invoker can be async (i.e. returns Task<object?>)
-            Func<object?[], Task<object>> invoker = async args =>
+            async Task<object?> Invoker(object?[] args)
             {
                 // Wrap in a scope (even for statics to keep logic simple)
-                using (var scope = provider.CreateScope())
+#if NET6_0_OR_GREATER
+                var scope = provider.CreateAsyncScope();
+#else
+                var scope = provider.CreateScope();
+#endif
+                try
                 {
-                    object? result;
+                    object? result = null;
                     try
                     {
                         // Invoke static or non-static
-                        var instance = method.IsStatic ? null : scope.ServiceProvider.GetRequiredService(instanceType);
+                        var instance = method.IsStatic
+                            ? null
+                            : scope.ServiceProvider.GetRequiredService(instanceType);
+
                         result = method.Invoke(instance, args);
                     }
                     catch (TargetInvocationException e)
                     {
+#if NET6_0_OR_GREATER
                         ExceptionDispatchInfo.Capture(e.InnerException!).Throw();
+#else
+                        ExceptionDispatchInfo.Capture(e.InnerException).Throw();
                         // Unreachable
                         throw new InvalidOperationException("Unreachable");
+#endif
```

Comment:
```suggestion
#else
                        ExceptionDispatchInfo.Capture(e.InnerException).Throw();
#endif
                        // Unreachable
                        throw new InvalidOperationException("Unreachable");
```

I think this throw should be outside the conditional, and we should be able to remove ` = null` added to `object? result` above.

Reply:
So this was a bit of a stylistic choice.

In modern .NET the compiler knows that `.Throw()` from EDI is the end of the scope. Like `throw;` or `return;` and therefore it doesn't need the `throw new InvalidOperationException("")`. This was not true of the older compilers.

Using it as a preprocessor directive means that in the NET6, that second line (`throw new IOE()`) doesn't exist in the output MSIL. These can lead to (albeit micro) optimizations that the JITer can take advantage of from the simplified MSIL it's working with. The goal was to remove the `throw new IOE()` in NET6 . However, to keep it simple and clean, I would recommend just leaving them as this:

```csharp
ExceptionDispatchInfo.Capture(e.InnerException).Throw();
// Unreachable
throw new InvalidOperationException("Unreachable");
```

Reply:
:+1: The benefit for moving the unreachable part outside in this case is that now you don't have to initialize the variable because _all_ versions of .NET _know_ the variable is set before use. By doing version-specific things, you were forced to initialize a variable because not all versions could know it was set.